### PR TITLE
Allow withrawal of all funds

### PR DESCRIPTION
**Pull request by CodeSandwich**
Fixes https://github.com/radicle-dev/radicle-contracts/issues/55.

The `WITHDRAW_ALL` value is not `0`, because in the upcoming batching API we'll need to somehow convey that the user doesn't want to withdraw anything. Solidity doesn't make it easy to have optional arguments in the APIs, so the value `0` looks like the best bet to express that.

### CODE REVIEW

**Review by NunoAlexandre — COMMENTED** [review]
Thanks, Igor! :smiley: 

 Now, if we want to withdraw all the collectable funds, we'd need to first import `WITHDRAW_ALL` and pass that to mean "collect all funds".

I'd prefer to have a cleaner API as a client. Either:

1. Offer both `withdraw(uint128)` and `withdrawAll()`, with the latter calling the first with `WITHRAW_ALL` as an argument, freeing the client from that responsibility and making the API more ergonomic

2. Drop the `uint128` argument and just call `withdrawInternal` with `WITHDRAW_ALL`. We don't foresee any use case to support a partial withdraw so this solution is cleanest and checks all boxes.

File: test/pool.test.ts [hunk 1]
```diff
@@ -522,6 +522,25 @@ describe("EthPool", function () {
...
   });
+
+  it("Allows withdrawal of all funds", async function () {
+    const [sender] = await getEthPoolUsers();
+    const withdrawAll = await sender.pool.WITHDRAW_ALL();
+    const amount = 1234;
+    await sender.topUp(0, amount);
+    const balanceBefore = await sender.pool.signer.getBalance();
+    await submit(
+      sender.pool.withdraw(withdrawAll, {gasPrice: 0}),
```
Comment (geigerzaehler) [inline test/pool.test.ts:533]: I’d feel more comfortable with the test if we also included some receivers to cover a reduction in the balance. However, if this is ok for you we can also leave as it is.

**Review by geigerzaehler — APPROVED** [review]

**Comment by CodeSandwich** [comment]
> I'd prefer to have a cleaner API as a client.

I've been considering 1, but it's not going to go well with batching. If we can withdraw some or withdraw all or withdraw nothing, we'll have to start juggling with enums and optional parameters. We need an API, which can transform straight to the batching one.

I'm not convinced that we want to drop partial withdrawals, we were discussing this during the last sync and to me it seemed like we all agreed that this elasticity is important for the users. If we drop it, we'll force users to double time and cost if they want to withdraw, but not everything.

**Comment by NunoAlexandre** [comment]
> > I'd prefer to have a cleaner API as a client.
> 
> I've been considering 1, but it's not going to go well with batching. If we can withdraw some or withdraw all or withdraw nothing, we'll have to start juggling with enums and optional (ignored) parameters. We need an API, which can transform straight to the batching one.
> 
> I'm not convinced that we want to drop partial withdrawals, we were discussing this during the last sync and to me it seemed like we all agreed that this elasticity is important for the users. If we drop it, we'll force users to double time and cost if they want to withdraw, but not everything.

Can you explain how this withdrawing feature touches the batching feature? The latter should only involve setting the receivers, the monthly contribution, and the initial balance (top up). Am I missing something? :thinking:

**Comment by CodeSandwich** [comment]
Sorry, batching and batching :D I mean the API which allows you to set up or update the whole sender in one shot.

**Review by NunoAlexandre — APPROVED** [review]
Let's go as-is :surfer:

File: test/pool.test.ts (hunk 1 above)
Comment (CodeSandwich) [reply test/pool.test.ts:535]: Fixed

**Comment by NunoAlexandre** [comment]
> I'm not convinced that we want to drop partial withdrawals, we were discussing this during the last sync and to me it seemed like we all agreed that this elasticity is important for the users. If we drop it, we'll force users to double time and cost if they want to withdraw, but not everything.

For the record: sorry, I was confusing this with the collect funds feature :see_no_evil: good that you didn't.

**Pull request merged**